python/ray/_raylet.pyx

```diff
@@ -1518,6 +1529,203 @@ cdef class EmptyProfileEvent:
     def __exit__(self, *args):
         pass
 
+
+class RetryGcsConnectionError(Exception):
+    "If this error is raised, we should try to reconnect to GCS."
+    def __init__(self, exception):
+        self.exception = exception
+
+
+def _auto_reconnect(f):
+    @wraps(f)
+    def wrapper(self, *args, **kwargs):
+        if "TEST_RAY_COLLECT_KV_FREQUENCY" in os.environ:
+            ray._private.utils._CALLED_FREQ[f.__name__] += 1
+        remaining_retry = self._nums_reconnect_retry
+        while True:
+            try:
+                return f(self, *args, **kwargs)
+            except RetryGcsConnectionError as e:
+                if remaining_retry <= 0:
+                    raise e.exception
+                logger.debug(
+                    f"Failed to send request to gcs, reconnecting. Error {e.exception}"
+                )
+                try:
+                    self._connect()
+                except Exception:
+                    logger.error(f"Connecting to gcs failed. Error {e.exception}")
+                time.sleep(1)
+                remaining_retry -= 1
+                continue
+    return wrapper
+
+
+cdef class GcsClient:
+    """Cython wrapper class of C++ `ray::gcs::GcsClient`."""
+    cdef:
+        shared_ptr[CGcsSyncClient] inner
+        object address
+        object _nums_reconnect_retry
+
+    def __cinit__(self, address, nums_reconnect_retry=5):
+        cdef GcsClientOptions gcs_options = GcsClientOptions.from_gcs_address(address)
+        self.inner.reset(new CGcsSyncClient(dereference(gcs_options.native())))
+        self.address = address
+        self._nums_reconnect_retry = nums_reconnect_retry
+        self._connect()
+
+    def _connect(self):
+        check_status(self.inner.get().Connect())
+
+    cdef _check_error(self, CRayStatus status):
+        try:
+            check_status(status)
+        except RaySystemError as e:
+            if status.IsGrpcUnavailable() or status.IsGrpcUnknown():
+                raise RetryGcsConnectionError(e)
+            else:
+                raise RuntimeError(str(e))
+
+    @property
+    def address(self):
+        return self.address
+
+    @property
+    def _nums_reconnect_retry(self):
+        return self._nums_reconnect_retry
+
+    @_auto_reconnect
+    def internal_kv_get(self, bytes key, namespace=None, timeout=None):
```

Comment:
would it be better if we allow retry number on per-call invocation? (if so, I guess this could also be in a follow-up)

Reply:
oh, i see this is the same interface we are exposing right now.